### ISSUE TITLE
Problem with Image2Image in the test phase

### ISSUE THREAD

**Issue by kaihuchen** [ppwwyyxx/tensorpack#80]
@ppwwyyxx 

I am testing the Image2Image component and have no problem getting it to train on 400x400 images. However, when I tested it with another batch of 400x400 (or 800x400 in the AB format) images the program reports the following error:

    ValueError: Cannot feed value of shape (6, 400, 400, 3) for Tensor u'input:0', which has shape '(?, 256, 256, 3)' 

so I just resize the test images to 516x256 (in AB format), but then got another error as follows:

<pre>
Restoring checkpoint from ./train_log/Image2Image/model-last.index ...
  0%|                                                                                                       |0/1[00:00<?,?it/s]
Traceback (most recent call last):
  File "./Image2Image.py", line 202, in <module>
    sample(args.data, args.load)
  File "./Image2Image.py", line 186, in sample
    for o in pred.get_result():
  File "/mnt2/mllocal/tensorpack/tensorpack/predict/dataset.py", line 65, in get_result
    res = self.predictor(dp)
  File "/mnt2/mllocal/tensorpack/tensorpack/predict/base.py", line 36, in __call__
    output = self._do_call(dp)
  File "/mnt2/mllocal/tensorpack/tensorpack/predict/base.py", line 82, in _do_call
    output = self.session.run(self.output_tensors, feed_dict=feed)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/client/session.py", line 766, in run
    run_metadata_ptr)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/client/session.py", line 964, in _run
    feed_dict_string, options, run_metadata)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/client/session.py", line 1014, in _do_run
    target_list, options, run_metadata)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/client/session.py", line 1034, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.FailedPreconditionError: Attempting to use uninitialized value gen/deconv8/b
         [[Node: gen/deconv8/b/read = Identity[T=DT_FLOAT, _class=["loc:@gen/deconv8/b"], _device="/job:localhost/replica:0/task:0/cpu:0"](gen/deconv8/b)]]

Caused by op u'gen/deconv8/b/read', defined at:
  File "./Image2Image.py", line 202, in <module>
    sample(args.data, args.load)
  File "./Image2Image.py", line 185, in sample
    pred = SimpleDatasetPredictor(pred, ds)
  File "/mnt2/mllocal/tensorpack/tensorpack/predict/dataset.py", line 54, in __init__
    self.predictor = OfflinePredictor(config)
  File "/mnt2/mllocal/tensorpack/tensorpack/predict/base.py", line 93, in __init__
    config.model.build_graph(input_placehdrs)
  File "/mnt2/mllocal/tensorpack/tensorpack/models/model_desc.py", line 88, in build_graph
    self._build_graph(model_inputs)
  File "./Image2Image.py", line 106, in _build_graph
    fake_output = self.generator(input)
  File "./Image2Image.py", line 81, in generator
    .Deconv2D('deconv8', OUT_CH, nl=tf.tanh)())
  File "/mnt2/mllocal/tensorpack/tensorpack/models/__init__.py", line 60, in f
    ret = layer(name, self._t, *args, **kwargs)
  File "/mnt2/mllocal/tensorpack/tensorpack/models/_common.py", line 75, in wrapped_func
    outputs = func(*args, **actual_args)
  File "/mnt2/mllocal/tensorpack/tensorpack/models/conv2d.py", line 127, in Deconv2D
    b = tf.get_variable('b', [out_channel], initializer=b_init)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/variable_scope.py", line 1024, in get_variable
    custom_getter=custom_getter)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/variable_scope.py", line 850, in get_variable
    custom_getter=custom_getter)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/variable_scope.py", line 346, in get_variable
    validate_shape=validate_shape)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/variable_scope.py", line 331, in _true_getter
    caching_device=caching_device, validate_shape=validate_shape)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/variable_scope.py", line 677, in _get_single_variable
    expected_shape=shape)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/variables.py", line 224, in __init__
    expected_shape=expected_shape)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/variables.py", line 370, in _init_from_args
    self._snapshot = array_ops.identity(self._variable, name="read")
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/gen_array_ops.py", line 1424, in identity
    result = _op_def_lib.apply_op("Identity", input=input, name=name)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/op_def_library.py", line 759, in apply_op
    op_def=op_def)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/ops.py", line 2240, in create_op
    original_op=self._default_original_op, op_def=op_def)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/ops.py", line 1128, in __init__
    self._traceback = _extract_stack()

FailedPreconditionError (see above for traceback): Attempting to use uninitialized value gen/deconv8/b
         [[Node: gen/deconv8/b/read = Identity[T=DT_FLOAT, _class=["loc:@gen/deconv8/b"], _device="/job:localhost/replica:0/task:0/cpu:0"](gen/deconv8/b)]]
</pre>

Any suggestion regarding the best way to get over this problem? Thanks!

**Comment by ppwwyyxx** [comment]
1. This script assumes images are all 256x256. In training, all training images are resized to 286 and cropped to 256 (`get_data()`) following the exact setup in the paper. Therefore to use the trained model, you'll need to resize the images, or retrain with 400x400 resolution.

2. For Tensorflow SaverV2 format, you'll need to pass `/path/to/model-last` to the saver, instead of the `.index` or `.data` file. (See #79). 
I don't like this feature of Tensorflow as well, so in one recent commit (f1fdb42e2f3d8), the loader in tensorpack now automatically replace `model-last.index` with `model-last`. If you update the code it should work fine.

**Comment by kaihuchen** [comment]
@ppwwyyxx 
Thanks for the quick reply. I have been using pix2pix/Torch for while, in fact all of the 400x400 training/testing datasets I used here were borrowed directly from my pix2pix/Torch. Now that I know that tensorpack requires 256x256 images I have adjusted my datasets accordingly. After adjusting for the model path as you mentioned above, I got the following error when running the test samples:

<pre>
[0101 00:14:08 @sessinit.py:71] Restoring checkpoint from ./train_log/Image2Image0101-000755/model-18 ...
[0101 00:14:08 @sessinit.py:141] WRN Variable pos_acc/EMA/local_step:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable input_queue_size/EMA:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable global_step:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable total_g_loss/EMA/biased:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable d_CE_loss_neg/EMA:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable g_CE_loss/EMA/biased:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable g_CE_loss/EMA:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable g_CE_loss/EMA/local_step:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable total_g_loss/EMA/local_step:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable d_CE_loss_pos/EMA/biased:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable neg_acc/EMA/local_step:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable d_CE_loss/EMA:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable pos_acc/EMA:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable beta1_power:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable input_queue_size/EMA/biased:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable L1_loss/EMA:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable d_CE_loss_neg/EMA/local_step:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable L1_loss/EMA/biased:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable learning_rate:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable total_g_loss/EMA:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable d_CE_loss/EMA/local_step:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable neg_acc/EMA/biased:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable beta2_power:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable neg_acc/EMA:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable L1_loss/EMA/local_step:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable pos_acc/EMA/biased:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable input_queue_size/EMA/local_step:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable d_CE_loss/EMA/biased:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable d_CE_loss_pos/EMA:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable d_CE_loss_pos/EMA/local_step:0 in checkpoint not found in the graph!
[0101 00:14:08 @sessinit.py:141] WRN Variable d_CE_loss_neg/EMA/biased:0 in checkpoint not found in the graph!
  0%|                                                                                                                      |0/1[00:00<?,?it/s]Failed to connect to Mir: Failed to connect to server socket: No such file or directory
Unable to init server: Could not connect: Connection refused

(random_window_name:20692): Gtk-WARNING **: cannot open display:
</pre>

The error seems to be about some GTK display setup which I could do without. Is there any way to just see the output images as files, or else see them in the Tensorboard?

**Comment by ppwwyyxx** [comment]
Of course. In line 187 of Image2Image.py:
```python
        o = o[0][:,:,:,::-1]
```
`o` would be a tensor of shape `(6, h, w * 3, 3)` containing 6 RGB images. Each image is like the visualization I put in tensorboard: it's (input, ground truth, output).

You can delete the line`viz = next(build_patch_list(o, nr_row=3, nr_col=2, viz=True))` which opens a window for visualization, and save the 6 images manually to somewhere.

**Comment by kaihuchen** [comment]
@ppwwyyxx   
That's very helpful info which got me over the problem I had earlier. There is only one hitch remains where the images I saved look off-color, like this:

![1-out-0](https://cloud.githubusercontent.com/assets/13685803/21583150/7043d68a-d041-11e6-92db-faee3044c0d4.jpg)

The above is expected to be in full color instead. My code looks like this:

    for i,img in enumerate(o):
        img = Image.fromarray(img, "RGB")
        path = os.path.join("somePathHere-"+str(i)+".jpg")
        img.save(path)

I compared the pixel value in the array **img** with what got rendered in the file which appear to be the same, so I believe I am rendering exactly what got passed in **o**. Is there some pre-processing that I need to do with **img** here?  Thanks!
